subprojects/base-services/src/main/java/org/gradle/internal/Either.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal;
+
+import java.util.NoSuchElementException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+public abstract class Either<L, R> {
```

Comment:
Can we get rid of the either type? We currently don't use it.

 I also think it is not very good to expose the Either type on public method of a class, since `left` or `right` are not very good for expressing what is going on. It is very similar to `Pair`, which also can be used in an implementation, but shouldn't be exposed on an API. I think a visitor pattern tailored to the use-case is much better than using `Either`.

Reply:
👍 I removed `Either`. I thought it would be good, but in actual usage it turned out to be confusing.